File: 2022/6/6.py
#test1 = 'bvwbjplbgvbhsrlpgdmjqwftvncz'
#test2 = 'nppdvjthqldpwncqszvftbrmjlhg'
#test3 = 'nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg'

def buffer_scan(buffer):
    count = 0
    SOP_address = 0
    for index, char in enumerate(buffer):
        if count == 4:
            break
        slice = buffer[index : index + 4]
        for c in slice:
            if slice.count(c) > 1:
                count = 0
                break
            else : count += 1
            if count == 4:
                SOP_address = buffer.find(slice) + 4
                break
    return (SOP_address)

def buffer_scan_message(buffer):
    count = 0
    SOM_address = 0
    for index, char in enumerate(buffer):
        if count == 14:
            break
        slice = buffer[index : index + 14]
        for c in slice:
            if slice.count(c) > 1:
                count = 0
                break
            else : count += 1
            if count == 14:
                SOM_address = buffer.find(slice) + 14
                break
    return (SOM_address)

file = open('input.txt')
buffer = file.read()
print(buffer_scan(buffer))
print(buffer_scan_message(buffer))

File: 2022/6/6.js
function bufferScan(buffer) {
  let count = 0;
  let SOPAddress = 0;
  for (let index = 0; index < buffer.length; index++) {
    if (count === 4) {
      break;
    }
    const slice = buffer.slice(index, index + 4);
    for (const c of slice) {
      if (slice.split(c).length - 1 > 1) {
        count = 0;
        break;
      } else {
        count += 1;
      }
      if (count === 4) {
        SOPAddress = buffer.indexOf(slice) + 4;
        break;
      }
    }
  }
  return SOPAddress;
}

function bufferScanMessage(buffer) {
  let count = 0;
  let SOMAddress = 0;
  for (let index = 0; index < buffer.length; index++) {
    if (count === 14) {
      break;
    }
    const slice = buffer.slice(index, index + 14);
    for (const c of slice) {
      if (slice.split(c).length - 1 > 1) {
        count = 0;
        break;
      } else {
        count += 1;
      }
      if (count === 14) {
        SOMAddress = buffer.indexOf(slice) + 14;
        break;
      }
    }
  }
  return SOMAddress;
}

const fs = require("fs");
const buffer = fs.readFileSync("input.txt", "utf-8");
console.log(bufferScan(buffer));
console.log(bufferScanMessage(buffer));
